cpp/src/arrow/util/decimal.cc

```diff
@@ -742,6 +1106,78 @@ static inline uint64_t UInt64FromBigEndian(const uint8_t* bytes, int32_t length)
   return ::arrow::bit_util::FromBigEndian(result);
 }
 
+Result<Decimal32> Decimal32::FromBigEndian(const uint8_t* bytes, int32_t length) {
+  static constexpr int32_t kMinDecimalBytes = 1;
+  static constexpr int32_t kMaxDecimalBytes = 4;
+
+  if (ARROW_PREDICT_FALSE(length < kMinDecimalBytes || length > kMaxDecimalBytes)) {
+    return Status::Invalid("Length of byte array passed to Decimal32::FromBigEndian was ",
+                           length, ", but must be between ", kMinDecimalBytes, " and ",
+                           kMaxDecimalBytes);
+  }
+
+  const bool is_negative = static_cast<int8_t>(bytes[0]) < 0;
+
+  uint32_t result = 0;
```

Comment:
How about something like:
```suggestion
  int32_t result = is_negative ? 0xffffffff : 0;
```
which should automically get you the required sign extension?

Reply:
wouldn't that get lost by the `memcpy` right below it?

Reply:
Ah! it would only get lost if `kMaxDecimalBytes == length` in which case we were just copying all the bytes anyways and the sign extension would already be there. Good point!